### PR TITLE
Minimum Viable TWAP

### PR DESCRIPTION
**Pull request by dapp-whisperer**
# Description

See Alex's implementation at https://github.com/GalloDaSballo/twap-study
@GalloDaSballo any more notes to add

The purpose of the TWAP is to have a trailing average of the eBTC totalDebt value
This value is used in redemptions to determine the rate. The more that gets redeemed relative to the totalDebt, the higher the rate goes

(someone could use flashloans to inflate the debt temporarily and push the redemption fee down)

let's imagine a theoretical worst-case where the fee stays flat at 0.5% - then we've got a constant-sum AMM

of course that wouldn't happen but this is designed to reduce spot manipulation of the totalDebt value

in this case, we now take the lower of the TWAP value and the spot value. (so the fee scales faster than spot, in the case of the supply going up "naturally" or via flashloan)

## Known Issues
Until observed once, no redemptions are possible due to:

### Zero TWAP
When no time passes or observation has happened, it  leads to a div0 error, bricking redemptions.
In our case it’s considered the specification of the system that redemptions won’t be on for a while, so this is ok.
This is easy to resolve manually by taking a second CDP action or manually observing after some time.

> question: use totalDebt if TWAP reports zero? should we guard against this? @rayeaster @GalloDaSballo 
we could for example throw out a zero value and use spot. or just accept it.

example fix:
```solidity
function _openTestCDP(address _user, uint256 _coll, uint256 _debt) internal returns (bytes32) {
        dealCollateral(_user, _coll);
        vm.startPrank(_user);
        collateral.approve(address(borrowerOperations), type(uint256).max);
        bytes32 _cdpId = borrowerOperations.openCdp(_debt, bytes32(0), bytes32(0), _coll);
        vm.stopPrank();

        // TWAP must be non-zero for redemptions at system start
        vm.warp(block.timestamp + 1);
        activePool.update();

        return _cdpId;
    }
```
(note that this was reverted for final testing PR - all legacy tests advance the TWAP to equal spot)

# Test Fix Notes
### Echidna to Foundry Tests
These sequences are based on the old assumption of using the spot totalDebt.

There is now a convenience function in BaseFixture and corresponding Echdina setup
```solidity
function _syncSystemDebtTwapToSpotValue() internal {
        vm.warp(block.timestamp + activePool.PERIOD());
        activePool.update();
    }
```

note that each setUp() empirically resets timestamp at `1`

TODO: How to dedup? `BaseStorageVariables -> BaseFunctions`

### Broken Tests
Some tests remain broken due to a new issue I haven't explored yet.

## Additional Tests
- Implement Alex's tests from https://github.com/GalloDaSballo/twap-study

### CODE REVIEW

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        uint256 futureWeight = block.timestamp - data.t0;
+        uint256 maxWeight = PERIOD;
+
+        if (futureWeight > maxWeight) {
+            _update(virtualAvgValue, priceCum0, uint64(block.timestamp)); // May as well update
+            // Return virtual
+            return virtualAvgValue;
+        }
+
+        uint256 weightedAvg = data.avgValue * (maxWeight - futureWeight);
```
[the remark this answers is not in the document]
Comment (rayeaster) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:103]: Original impl should be good according to @CodingNameKiki comment. https://github.com/ebtc-protocol/ebtc/pull/736#issuecomment-1853590072

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+pragma solidity 0.8.17;
+import {ITwapWeightedObserver} from "../Interfaces/ITwapWeightedObserver.sol";
+
+contract TwapWeightedObserver is ITwapWeightedObserver {
+    PackedData public data;
+    uint128 public valueToTrack;
+
+    constructor(uint128 initialValue) {
+        PackedData memory cachedData = PackedData({
+            priceCumulative0: initialValue,
```
[the remark this answers is not in the document]
Comment (dapp-whisperer) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:11]: agree on more generic naming

File: packages/contracts/contracts/Interfaces/ITwapWeightedObserver.sol [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License Identifier: MIT
+pragma solidity 0.8.17;
+import {IBaseTwapWeightedObserver} from "./IBaseTwapWeightedObserver.sol";
+
+interface ITwapWeightedObserver is IBaseTwapWeightedObserver {
+    function PERIOD() external view returns (uint256);
+
+    function valueToTrack() external view returns (uint128);
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Interfaces/ITwapWeightedObserver.sol:8]: Yes, removed

File: packages/contracts/contracts/Interfaces/ITwapWeightedObserver.sol [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    function timeToAccrue() external view returns (uint64);
+
+    function getRealValue() external view returns (uint256);
+
+    function getLatestAccumulator() external view returns (uint128);
+
+    // function setValue(uint128 newValue) external;
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Interfaces/ITwapWeightedObserver.sol:16]: Fixed

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+            _update(avgValue, priceCum0, time0);
+        }
+    }
+
+    /// Internal update so we can call it both in _update and in observe
+    function _update(uint128 avgValue, uint128 priceCum0, uint64 time0) internal {
+        data.avgValue = avgValue;
+        data.priceCumulative0 = priceCum0;
+        data.t0 = time0;
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:130]: Please send this as PR

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+        uint256 weightedAvg = data.avgValue * (maxWeight - futureWeight);
+        uint256 weightedVirtual = virtualAvgValue * (futureWeight);
+
+        uint256 weightedMean = (weightedAvg + weightedVirtual) / PERIOD;
+
+        return weightedMean;
+    }
+
+    function update() public {
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:111]: Am not convinced this is ever used, but lmk if you find a use case

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        // Let's virtally sync TWAP
+        // With a weight, that is higher, the more time has passed
+        uint128 priceCum0 = getLatestAccumulator();
+        uint128 virtualAvgValue = (priceCum0 - data.priceCumulative0) /
+            (uint64(block.timestamp) - data.t0);
+
+        uint256 futureWeight = block.timestamp - data.t0;
+        uint256 maxWeight = PERIOD;
+
+        if (futureWeight > maxWeight) {
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:97]: Don't disagree, please send PR

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+        // A reference period is 7 days
+        // For each second passed after update
+        // Let's virtally sync TWAP
+        // With a weight, that is higher, the more time has passed
+        uint128 priceCum0 = getLatestAccumulator();
+        uint128 virtualAvgValue = (priceCum0 - data.priceCumulative0) /
+            (uint64(block.timestamp) - data.t0);
+
+        uint256 futureWeight = block.timestamp - data.t0;
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:94]: Agree fixed

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        if (block.timestamp - data.t0 == 0) {
+            return data.avgValue;
+        }
+
+        // A reference period is 7 days
+        // For each second passed after update
+        // Let's virtally sync TWAP
+        // With a weight, that is higher, the more time has passed
+        uint128 priceCum0 = getLatestAccumulator();
+        uint128 virtualAvgValue = (priceCum0 - data.priceCumulative0) /
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:91]: Please send PR

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+contract TwapWeightedObserver is ITwapWeightedObserver {
+    PackedData public data;
+    uint128 public valueToTrack;
+
+    constructor(uint128 initialValue) {
+        PackedData memory cachedData = PackedData({
+            priceCumulative0: initialValue,
+            accumulator: initialValue,
+            t0: uint64(block.timestamp),
+            lastUpdate: uint64(block.timestamp),
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:14]: No opinion

File: packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol [hunk 11]
```diff
@@ -0,0 +1,134 @@
+// SPDX-License Identifier: MIT
+pragma solidity 0.8.17;
+import {ITwapWeightedObserver} from "../Interfaces/ITwapWeightedObserver.sol";
+
+contract TwapWeightedObserver is ITwapWeightedObserver {
```
[the remark this answers is not in the document]
Comment (GalloDaSballo) [reply packages/contracts/contracts/Dependencies/TwapWeightedObserver.sol:5]: Please give it a first pass

**Comment by GalloDaSballo** [comment]
@rayeaster please see my renaming and added coments

**Pull request merged**